toodlext/src/main.rs

```diff
@@ -0,0 +1,355 @@
+    }
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(tag = "type")]
+enum RequestBody {
+    CreateTodo(ItemInfo),
+    GetTodos,
+    TodoChangeName { uuid: String, name: String },
+    TodoChangeDueDate {
+        uuid: String,
+
+        #[serde(rename = "dueDate")]
+        due_date: i64,
+    },
+    TodoChangeCompletionDate {
+        uuid: String,
+
+        #[serde(rename = "completionDate")]
+        completion_date: i64,
+    },
+    TodoAddLabel { uuid: String, label: LabelInfo },
+    TodoRemoveLabel { uuid: String, name: String },
+    RemoveTodo { uuid: String },
+    AddLabel(LabelInfo),
+    RemoveLabel { name: String },
+    GetLabels,
+}
+
+#[derive(Deserialize, Debug)]
+struct Request {
+    id: i64,
+    body: RequestBody,
+}
+
+impl Request {
+    fn read_from(input: &mut StdinLock) -> Result<Request, Error> {
+        match input.read_u32::<NativeEndian>() {
+            Ok(length) => {
+                let mut message = input.take(length as u64);
+                let mut buffer = Vec::with_capacity(length as usize);
+                eprintln!("Reading request from browser");
+                match message.read_to_end(&mut buffer) {
+                    Ok(_) => {
+                        serde_json::from_slice(&buffer).map_err(|err| {
+                            eprintln!("Error parsing request payload {:?}: {:?}",
+                                      String::from_utf8_lossy(&buffer),
+                                      err);
+                            Error::BadJSON
+                        })
+                    }
+                    _ => Err(Error::IOError),
+                }
+            }
+            _ => Err(Error::IOError),
+        }
+    }
+}
+
+#[derive(Serialize, Debug)]
+struct Response {
+    id: i64,
+    ok: bool,
+    body: ResponseBody,
+}
+
+impl Response {
+    fn write_to(&self, output: &mut StdoutLock) -> io::Result<()> {
+        let message = serde_json::to_vec(self)?;
+        output.write_u32::<NativeEndian>(message.len() as u32)?;
+        output.write_all(&message)?;
+        output.flush()
+    }
+}
+
+#[derive(Serialize, Debug)]
+#[serde(untagged)]
+enum ResponseBody {
+    CreateTodo(ItemInfo),
+    UpdateTodo(ItemInfo),
+    GetTodos(Vec<ItemInfo>),
+    GetLabels(Vec<LabelInfo>),
+    RemoveTodo { uuid: String },
+    AddLabel(LabelInfo),
+    RemoveLabel { name: String },
+    Err(Error),
+}
+
+fn update_item_by_uuid(toodle: &mut Toodle,
+                       uuid: &str,
+                       name: Option<String>,
+                       due_date: Option<Timespec>,
+                       completion_date: Option<Timespec>)
+                       -> ResponseBody {
+    if let Some(uuid) = Uuid::parse_str(&uuid).ok() {
+        if let Some(ref item) = toodle.fetch_item(&uuid).ok().unwrap_or_default() {
+            if let Some(ref new_item) =
+                toodle
+                    .update_item(item, name, due_date, completion_date, None)
+                    .ok()
```

Comment:
Do you need this `ok` here? `and_then` should accept the value inside the `Result` as an arg, not the result itself - unless of course the result of `update_item` is a `Result` inside a `Result`, which I don't think it is.